mlir/lib/Catalyst/Transforms/QnodeToAsyncPatterns.cpp

```diff
@@ -31,6 +31,135 @@ namespace catalyst {
+        //
+        //     %1 = async.await %promise // promise is not yet available, will wait.
+        //     %2 = async.await %promise // promise is now available, will not wait.
+        //
+        //  This makes the job of deciding where to place awaits really easy.
+        //  We just need to place awaits just before every use of the original value.
+        //  For example, take the following piece of code
+        //
+        //     %quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %some_val = some.op %quantum_results
+        //
+        //  First we translate the call to an async.execute (which was performed before this
+        //  function was called) When this function was called, the transformation is not yet
+        //  complete. Because there are still uses of %quantum_results.
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %some_val = some.op %quantum_results
+        //
+        //  This function will perform the following change:
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %await_quantum_results = async.await %async_quantum_results
+        //     %some_val = some.op %await_quantum_results
+        //
+        //  And it will do so for every use of %quantum_results
+        //  E.g.
+        //
+        //     %async_quantum_results = call @qnode
+        //
+        //     // Arbitrary control flow
+        //     // This use might be inside a different basic block.
+        //     %await_quantum_results_0 = async.await %async_quantum_results
+        //     %some_val_0 = some.op %await_quantum_results_0
+        //     %await_quantum_results_1 = async.await %async_quantum_results_1
+        //     %some_val_1 = some.op %await_quantum_results_1
+        //
+        //  However, we can do a bit better and avoid this extra awaits that are
+        //  dominated by the first await. Keeping the awaits is not incur a large increase
+        //  in run time. Removing the awaits involves using the dominator analysis, which we need
+        //  some time to investigate how to use. Not much, but sufficient enough for a future
+        //  improvement. See TODO inside replaceUsesWithIf.
+        auto results = executeOp.getResults();
+        // The first value is just a token.
+        std::vector<Value> bodyReturns(results.begin() + 1, results.end());
+        // TODO: Assert/FAIL that op.getResults() and bodyReturns have the same number of elements.
+        if (bodyReturns.size() == op.getResults().size()) {
+            for (auto [oldVal, newVal] : llvm::zip(op.getResults(), bodyReturns)) {
```

Comment:
With double refs, you can advantage from both rvalue and lvalue as well as preserving the constness:

 ```suggestion
             for (auto &&[oldVal, newVal] : llvm::zip(op.getResults(), bodyReturns)) {
```